src/cosalib/meta.py

```diff
@@ -1,15 +1,86 @@
 import json
 import jsonschema
 import os.path
+import time
 
 from cosalib.builds import Builds
-from cosalib.cmdlib import write_json
+from cosalib.cmdlib import (
+    merge_dicts,
+    load_json,
+    write_json
+)
 
+COSA_DELAYED_MERGE = "coreos-assembler.delayed-meta-merge"
+COSA_VER_STAMP = "coreos-assembler.meta-stamp"
 
 SCHEMA_PATH = os.environ.get("COSA_META_SCHEMA",
                              "/usr/lib/coreos-assembler/schema/v1.json")
 
 
+class COSAMergeError(Exception):
+    """
+    Raised when unable to merge meta.json together
+    """
+    def __str__(self):
+        return f"unable to merge: {': '.join(self.args)}"
+
+
+def merge_meta(x, y):
+    """
+    Merge two GenericMeta's together
+
+    This is likely over engineered, merge_dicts is safe enough.
+    But to we want to be sure.
+
+    As a reminder: merge_dicts only merges the superset
+
+    Rules for merging:
+    - the first dict should be on-disk while second is modified one
+    - if the attribute checked exists on x, it must match on y.
+    - which ever version is newer is merged into the older
+
+    Note: this will NOT deal with replacing keys. For example,
+          re-running an AWS replication will not update the missing
+          keys.
+
+    """
+
+    # Only allow merging of artifacts from the same content
+    # i.e. you can't add RHCOS into FCOS.
+    for i in ["ostree-commit", "ostree-content-checksum",
+              "coreos-assembler.image-config-checksum"]:
+        if x.get(i, False) and (x.get(i) != y.get(i)):
+            raise COSAMergeError(f"meta.json conflict on '{i}' "
+                                 f"{x.get(i)} != {y.get(i)}")
+
+    # Do not allow merging from different COSA code.
```

Comment:
Hmm, do we really want that? One of the main end goals is to be able to rerun parts of the pipeline that failed, and it's possible that the failure will need a cosa fix.

Reply:
This is tricky, and I think erroring on the side of caution here is warranted. A "build" is a composite of the input -- the source, the recipe, and the builder. Building different artifacts via mismatched COSA's is analogous to using a different version of GCC.

Reply:
Makes sense. I think we will need more flexibility in the future re. this. But we can broaden it when we get there.